RFC: Bluetooth: Fix building applications with Privacy without ACL or SMP

cvinayak opened the pull request:
Update Host implementation so that Host can use the Controller Privacy feature to generate Advertising Resolvable Private
Address without enabling ACL Connection and Security Manager Protocol support in application builds.

Fix building Controller with Privacy feature without enabling LE Encryption support. This should allow being either Observer or Broadcaster role with Privacy features.

From https://github.com/zephyrproject-rtos/zephyr/pull/89408#discussion_r2129087202
As a developer of an LE Audio Auracast transmitter would like to use unique Identity Root and unique Identity Address in every nRF SoC to advertise using resolvable private address. My Auracast receivers (plural) are programmed with my transmitter's unique Identity Address and Identity Resolution Key as peer Id address and peer IRK, respectively, such that using filter accept list will enable the receiver's Controller implement to perform periodic advertising sync without the Host needing to receive redundant advertising reports of other devices in the vicinity.

cvinayak committed:
Bluetooth: Controller: Fix building with Privacy without LE Encryption

Fix building with Privacy feature without enabling LE
Encryption support. This should allow being either Observer
or Broadcaster role with Privacy features.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

omkar3141 commented:
Will there be any system test added for this?

Thalley reviewed (changes requested):
Overall the changes look OK, but please expand your commit message on what the issue is/was and how this fixes it: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-body

cvinayak committed:
tests: bsim: Bluetooth: Conditionally compile Controller Privacy

Conditionally compile Controller Privacy testing and enable
advertising test for nRF54L15bsim.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Host: Support Controller Privacy without ACL Connections

Update implementation so that Host can use the Controller
Privacy feature to generate Advertising Resolvable Private
Address without enabling ACL Connection and Security Manager
Protocol support in application builds.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (commented):
Overall the changes look OK to me, but there is some mismatches on some guards it looks like

cvinayak commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -296,23 +296,26 @@ struct bt_dev_le {
 
-#if defined(CONFIG_BT_SMP)
+#if defined(CONFIG_BT_PRIVACY)
 	/* Size of the controller resolving list */
 	uint8_t                    rl_size;
 	/* Number of entries in the resolving list. rl_entries > rl_size
 	 * means that host-side resolving is used.
 	 */
 	uint8_t                    rl_entries;
-#endif /* CONFIG_BT_SMP */
+#endif /* CONFIG_BT_PRIVACY */
```
![image](https://github.com/user-attachments/assets/0373f92f-f813-4c2d-bc30-05a39da09a24)

> Should this also be guarded by CONFIG_BT_SMP?

There was a request to meet Broadcaster Privacy Feature, wherein it is brought to notice that creating new identity did not lead to advertising with use of a resolvable private address using this new identity (may be this is handled correctly as Host privacy, but current Host does not use Controller privacy available). 

For enabling Controller Privacy (i.e. Controller generated resolvable private address) in Broadcaster-only builds, the resolving list will have to be populated with the local IRK. Likewise, if Observer uses filter list with Controller private address resolution then both id address and peer IRK in the resolving list is needed,

Hence, resolving list needs to be decoupled from CONFIG_BT_SMP feature.

> Some places in this commit you've guarded access to resolving list (and the related HCI commands) with both Kconfig options, and some places with just CONFIG_BT_PRIVACY.

Currently, resolving list population is only done with SMP feature, hence had to guard with both. But once I add samples/tests for broadcaster-only and observer-only, I will clean up the implementation on a test-driven basis. This is WIP is progress, will put this PR in draft (but solicit early comments and change requests during this time).

> Is the resolving list a feature that should be supported without SMP?

@omkar3141 is it possible to make a Broadcaster Only device design listing with the above 11/2 item selected but without SMP ? can you confirm?

Thalley replied:
Does privacy only apply to centrals, peripherals and broadcasters? 
i.e. is it a feature that can also be used by observers only?

omkar3141 replied:
> is it possible to make a Broadcaster Only device design listing with the above 11/2 item selected but without SMP ? can you confirm?

Going by the GAP ICS. GAP 11/2 "Resolvable private address generation procedure [5] 10.8.2.2" is mandatory if GAP 11/1 "Privacy feature" is supported. Now the [Privacy Feature](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-60/out/en/host/generic-access-profile.html#UUID-9e974725-ff34-67ae-0899-330726b7b296) further states that:

"A Peripheral shall use non-resolvable or resolvable private addresses when in non-connectable mode as defined in Section 9.3.2."

So, according to this, it is possible to make a peripheral-broadcaster product, that simply advertises non-connectable ADVs with RPAs. Now, as such a product never establishes a connection, nothing in SMP is needed. This is also reflected in GAP ICS as there is no mandatory linkage between SMP and RPA generation (or resolution).

cvinayak replied:
No, this is where i am stuck :-( ... comments are welcome.

PavelVPV replied:
Maybe we could add an API that takes local identity handle, peer address and peer IRK as arguments and then does what `bt_id_add` and `bt_id_delete` do except the part that changes `struct bt_keys::state` which should be compiled in when `BT_SMP` is enabled. But it is not clear for me in these use-cases how device gets peer address and peer IRK. And maybe the scope is even bigger than that.

Thalley replied:
> But it is not clear for me in these use-cases how device gets peer address and peer IRK. And maybe the scope is even bigger than that.

Imagine a feature where if you are logged in on both your phone and PC with some account, then pairing a device on your phone may share the remote keys with your PC via some online cloud service, and the device is then automatically bonded on your PC (assuming that the PC and Phone somehow share the same IRK and address when connecting to devices). IIRC Apple devices have such a feature, and such and API would allow users to implement the same functionality on Zephyr devices

cvinayak commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -295,6 +295,101 @@ config BT_FILTER_ACCEPT_LIST
+	  advertiser is only possible if the controller also supports privacy.
+
+config BT_PRIVACY_RANDOMIZE_IR
+	bool "Randomize identity root for fallback identities"
```
![image](https://github.com/user-attachments/assets/8a873ebc-9690-4412-bc46-fb8a3419f719)

Unfortunately, latest Core Spec lost this in the abbreviation section!  But references it still:
![image](https://github.com/user-attachments/assets/b620eca8-e061-42d0-bc4e-6d707e5af821)

PavelVPV commented on subsys/bluetooth/host/id.c:
```diff
@@ -1987,6 +2023,22 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
 #endif /* defined(CONFIG_BT_OBSERVER) */
+
+#if defined(CONFIG_BT_PRIVACY)
+		} else if (BT_FEAT_LE_PRIVACY(bt_dev.le.features) && bt_dev.le.rl_entries &&
```
By `this` I meant line 2028, where `rl_entries` is checked.

PavelVPV replied:
But `rl_entries` is not changed if `BT_SMP` is disabled. So if only `BT_PRIVACY` is enabled, this code will be enabled, but it will never work because `BT_SMP` needs to be enabled as well to change `rl_entries`.

cvinayak replied:
This is where we need public API to add to resolving list (without BT_CONN and BT_SMP support).

See how BSIM CI using Controller Privacy here using private functions here: 
https://github.com/zephyrproject-rtos/zephyr/pull/89408/files#diff-c745c2384557fb6b2984edb2c1eab58cc574fc41a30e82858eb28c2668ddd403R1734-R1766

cvinayak commented on tests/bsim/bluetooth/ll/advx/src/main.c:
```diff
@@ -1716,28 +1724,46 @@ static void test_scanx_main(void)
+		printk("Add to resolving list...");
+		bt_addr_le_t some_id_addr = {
+			.type = BT_ADDR_LE_RANDOM,
+			.a = {
+				.val = {0x78, 0x87, 0x78, 0x87, 0x78, 0x87}
+			}
+		};
+		uint8_t pirk[16] = {0x12, 0x21, 0x12, 0x21, 0x12, 0x21, 0x12, 0x21,
+				    0x12, 0x21, 0x12, 0x21, 0x12, 0x21, 0x12, 0x21};
+		uint8_t lirk[16] = {0xCD, 0xDC, 0xCD, 0xDC, 0xCD, 0xDC, 0xCD, 0xDC,
+				    0xCD, 0xDC, 0xCD, 0xDC, 0xCD, 0xDC, 0xCD, 0xDC};
+
+		/* some_id_addr with swapped peer IRK and local IRK */
+		err = ll_rl_add(&some_id_addr, lirk, pirk);
+		if (err) {
+			goto exit;
+		}
 
-	printk("Enable resolving list...");
-	err = ll_rl_enable(BT_HCI_ADDR_RES_ENABLE);
-	if (err) {
-		goto exit;
+		/* peer_id_addr with correct peer IRK and local IRK */
+		err = ll_rl_add(&peer_id_addr, pirk, lirk);
+		if (err) {
+			goto exit;
+		}
+		printk("success.\n");
+
+		printk("Enable resolving list...");
+		err = ll_rl_enable(BT_HCI_ADDR_RES_ENABLE);
+		if (err) {
+			goto exit;
+		}
+		printk("success.\n");
```
RFC: Need public API to add to Resolving List without ACL and SMP.
